### DISCUSSION TITLE
How to efficiently apply a rotation vector field to EBSD data?

### DISCUSSION THREAD

**Discussion by creuzige** [mtex-toolbox/mtex#1958] [Ask Anything] [answered]
I have some EBSD data taken near a fracture surface (low resolution example image below).  In the image you can see there's a substantial amount of rotation of the grains on the top and bottom of the image due to rotation of the normal direction as the sample thins.  The middle has not rotated due to symmetry.  

![Area1-IPFY-scaled2](https://github.com/mtex-toolbox/mtex/assets/10198749/d416fbee-03bb-4722-9654-9f4a493b2180)

I'd like to analyze the grain orientation relative to the normal direction (ND) surface, but that would require me to rotate each EBSD pixel to different amounts (i.e. near the middle there's no rotation, but on the top and bottom of the image there'd be about a 30° rotation. I think I've got some ideas (omega of fitEllipse?) on how to develop a rotation vector field of the appropriate rotations (like a quiver plot), so I imagine I'd have two matrices to work with:
EBSD_data=[X Y Orientation ...]
Rotation=[X Y Rotation]  

Is there an efficient way of applying the Rotation to each Orientation based on the X Y coordiantes? I could imagine doing things with for loops, but that's almost always the wrong way in matlab...

**Comment by kilir** [comment, marked as answer]
Hi Adam,

something like this ? 
```
mtexdata copper

ebsd = ebsd.gridify;

% some fantasy rotation based on x,y, coordinates
rot = rotation.byEuler(0,10* cos(ebsd(:).y/ max(ebsd(:).y)), 0  )

rot = reshape(rot,size(ebsd,1),[])

% apply rot to ori

ebsd_r = ebsd;
% rotate it
ebsd_r.rotations =  rot .* ebsd_r.rotations;


plot(ebsd('c'),ebsd('c').orientations)
nextAxis
plot(ebsd_r('c'),ebsd_r('c').orientations)
```

Cheers,
Rüdiger

**Reply by creuzige** [reply]
That does seem to work well.  Ran into two follow on questions:

1. Is there a way to apply it as a angle/axis pair?  I feel like that might be more intuitive, but when I tried I got the following errors:
```matlab
rot=rotation.byAxisAngle('axis',repmat(zvector,size(ebsd(:))),'angle', ...
 (10*degree*(ebsd(:).y - mean(ebsd(:).y) ) ) );
```
```
Error using axis2quat
first argument should be vector3d and second the angle

Error in rotation.byAxisAngle (line 17)
rot = rotation(axis2quat(v,omega));
```

My read is that the byAxisAngle script is expecting a single term, not an array.  But the byEuler script can handle multiple arrays.  Does that seem correct, or did I mess something up on the syntax?

2. Is there a way to get the magnitude of the rotation directly from the rotation array?  I was able to get it via the angle between rotated and unrotated ebsd plots, but I'm concerned that might also be kind of costly.  
```
figure;
plot(ebsd('c'),angle(ebsd_r('c').orientations,ebsd('c').orientations)/degree)
mtexColorbar
```

Thanks again!

**Reply by ralfHielscher** [reply]
Hi Adam,
```matlab
rot=rotation.byAxisAngle(zvector, 10*degree * (ebsd.y - mean(ebsd.y(:)) );
```
should do the job. `rot.angle` gives you the angle :smile: 

Ralf.
